File: 3-unconstrained-optimization/matlab/ex_steepest_descent.m
% ex_steepest_descent.m
% examples of the steepest-descent method on quadratic functions
% [reference] Section 12.2 in LNO
% [course] Session 8 - Unconstrained Optimization (2)
close all; clear; clc

example = 4;

switch example
    %----------------------------------------------------------------------
    case 1 % Example 12.1 in LNO
        Q = diag([1 5 25]);
        c = [-1;-1;-1];
        x0 = [0;0;0];
        n = 217;
    %----------------------------------------------------------------------
    case 2 % cond(Q) is large
        d = [1 25];
        s = 3535;
        Q = create_rand_pd_matrix(d,s);
        c = [-1;-1];
        x0 = [0;0];
        n = 25;
    %----------------------------------------------------------------------
    case 3 % cond(Q) is smallish
        d = [1 2];
        s = 3243;
        Q = create_rand_pd_matrix(d,s);
        c = [-1;-1];
        x0 = [0;0];
        n = 20;
    %----------------------------------------------------------------------
    case 4 % cond(Q) = 1
        Q = diag([5 5]);
        c = [-1;-1];
        x0 = [0;0];
        n = 5;
end

% optimal solution
x_opt = Q\c;

% check if 2D problem that can be plotted
if length(Q) == 2
    n2flag = true;
else
    n2flag = false;
end

% initialize figure (2D problem only)
plot_2d([],[],[],[],[],[],n2flag,'init')

% assign initial point
x = x0;

% initialize function value storage
f = nan(1,n);

% go through each iteration
for k = 1:n

    % store old value
    xold = x;

    % compute current function value
    if k < 25
        f(k) = x'*Q*x/2 - c'*x;
    end

    % steepest-descent direction
    p = -(Q*x - c);

    % norm of the gradient
    norm_g = norm(Q*x - c);
    disp(strcat("norm(g) = ",string(norm_g)))

    % exact line search
    alpha = (p'*p)/(p'*Q*p);

    % next point
    x = x + alpha*p;

    % plot iteration (2D problem only)
    plot_2d(x,xold,x_opt,f,Q,c,n2flag,'iter')

end

% plot contours and optimal point (2D problem only)
plot_2d(x,xold,x_opt,f,Q,c,n2flag,'final')

%--------------------------------------------------------------------------
% create a random positive-definite matrix with given eigenvalues
function A = create_rand_pd_matrix(d,s)

% set random seed
if ~isempty(s)
    rng(s)
end

% size of the matrix
n = length(d);

% get a random rotation matrix
[Q, ~] = qr(rand(n));

% apply the transformation
A = Q * diag(d) * transpose(Q);

end

%--------------------------------------------------------------------------
% plotting functions
% (not the main content)
function plot_2d(x,xold,x_opt,f,Q,c,n2flag,tasknum)

% only do stuff if a 2D problem
if ~n2flag
    return
end

% colors and other parameters
niceblue = [77, 121, 167]/255;
nicered = [225, 86, 86]/255;
nicegreen = [109, 195, 80]/255;
LineWidth = 1.5;
MarkerSize = 24;
FontSize = 12;
plotOpts = {'LineWidth',LineWidth,'MarkerSize',MarkerSize};

switch tasknum
    %----------------------------------------------------------------------
    case 'init' % initial point

    % initialize figure
    hf = figure; hf.Color = 'w'; hold on

    % labels
    xlabel('$x$','Interpreter','latex');
    ylabel('$y$','Interpreter','latex');

    % axis properties
    ha = gca; ha.XColor = 'k'; ha.YColor = 'k'; ha.Color = 'none';
    ha.LineWidth = 1; ha.FontSize = FontSize;
    axis equal

    %----------------------------------------------------------------------
    case 'iter'

    % plot line between old and new point
    plot([xold(1) x(1)],[xold(2) x(2)],'-',plotOpts{:},'Color',niceblue);


    % plot current point
    plot(xold(1),xold(2),'.',plotOpts{:},'Color',nicered);

    %----------------------------------------------------------------------
    case 'final'

    % create a good set of contour lines
    f(isnan(f)) = [];
    fmax = 15;
    f_ = linspace(max(f),fmax,40);
    f = unique(sort([f,f_]));

    % create grid
    N = 1000;
    x = linspace(-1.5,0.5,N);
    y = linspace(-1,0.5,N);
    [X,Y] = meshgrid(x,y);

    % evaluate quadratic function
    C = zeros(size(X));
    for k = 1:numel(X)
        x = [X(k);Y(k)];
        C(k) = x'*Q*x/2 - c'*x;
    end

    % create contour plot
    [~,hc] = contour(X,Y,C,f);
    hc.LineWidth = LineWidth;
    colormap(bone)

    % plot optimal point
    plot(x_opt(1),x_opt(2),'.',plotOpts{:},'Color',nicegreen);

    % axis equal to visualize condition number
    axis equal

end

end

File: 3-unconstrained-optimization/python/ex_steepest_descent.py
import numpy as np 
import matplotlib.pyplot as plt
from numpy.linalg import norm,solve,qr
import numpy.random as random

def create_rand_pd_matrix(d,s):
    
    # set random seed
    random.seed(s)
        
    # size of the matrix 
    n = len(d)
    
    # get a random rotation matrix
    Q,r = qr(random.random((n,n)))

    # apply the transformation
    d_ = np.diag(d)
    A = Q.dot(d_.dot(Q))
    
    return A
    

test = 2

if test == 1: # example from the book
    Q = np.diag([1,5,25])
    c = np.array([-1,-1,-1])
    x0 = np.array([0,0,0])
    n = 217
    
elif test == 2: # cond(Q) is large
    d = np.array([1,25])
    s = 3525
    Q = create_rand_pd_matrix(d,s)
    c = np.array([-1,-1])
    x0 = np.array([0,0])
    n = 20
  
elif test == 3: # cond(Q) is smallish
    d = np.array([1,2])
    s = 3243
    Q = create_rand_pd_matrix(d,s)
    c = np.array([-1,-1])
    x0 = np.array([0,0])
    n = 20
    
elif test == 4: # cond(Q) = 1
    Q = np.diag([5,5])
    c = np.array([-1,-1])
    x0 = np.array([0,0])
    n = 5
    
# optimal solution 
x_opt = solve(Q,c)

# check if this is a 2d problem for plotting
if len(Q) == 2:
    n2flag = True
else:
    n2flaf = False

# setup figure
if n2flag:
    fig,ax = plt.subplots(1)
    ax.set_xlabel('x',fontsize = 16)
    ax.set_ylabel('y',fontsize = 16)
    
# assign initial point
x = x0

# initialize
f = n*[None]

# go through each iteration
for k in range(n):
    
    # store old value
    xold = x
    
    # compute current function value
    if k < 20:
        f[k] = x.dot(Q.dot(x))/2 - c.dot(x)
        
    # steepest-descent direction
    p = -(Q.dot(x) - c)
    
    # norm of the gradient
    norm_g = norm(Q.dot(x) - c)
    print('norm(g) = {}'.format(norm_g))
    
    # exact line search 
    alpha = p.dot(p)/(p.dot(Q.dot(p)))
    
    # next point 
    x = x + alpha*p 

    # plot if 2 dimensional
    if n2flag:
        
        # plot line between old and new point
        ax.plot([xold[0],x[0]],[xold[1],x[1]],'-b',linewidth = 1)
        
        # plot current point
        ax.plot(xold[0],xold[1],'.r',markersize = 12)
        
        
# plot if it is a two dimensional problem
if n2flag:
    
    # create grid
    N = 1000
    
    x = np.linspace(-1.5,0.5,N)
    y = np.linspace(-1,0.5,N)
    
    X,Y = np.meshgrid(x,y)
    
    X = X.reshape([N*N,])
    Y = Y.reshape([N*N,])
    C = np.zeros((N*N,))
    
    for k in range(N*N):
        x = np.array([X[k],Y[k]])
        C[k] = x.dot(Q.dot(x))/2 - c.dot(x)
    
    # reshape
    X = np.reshape(X,(N,N),order = 'F')
    Y = np.reshape(Y,(N,N),order = 'F')
    C = np.reshape(C,(N,N),order = 'F')
    
    ax.contour(X,Y,C)
    
    ax.plot(x_opt[0],x_opt[1],'.g',markersize = 12)
    
    plt.show()